File: src/main/java/ankol/mod/merger/tools/Tools.java
package ankol.mod.merger.tools;

import ankol.mod.merger.exception.BusinessException;
import cn.hutool.core.io.FileUtil;
import cn.hutool.core.util.StrUtil;
import lombok.Getter;
import org.apache.commons.compress.archivers.zip.ZipArchiveEntry;
import org.apache.commons.compress.archivers.zip.ZipFile;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.*;
import java.util.stream.Stream;

public abstract class Tools {
    @Getter
    private static final String userDir = System.getProperty("user.dir");
    @Getter
    private static final String tempDir = System.getProperty("java.io.tmpdir");

    // 复用十六进制表，避免 String.format 带来的巨大开销
    private static final char[] HEX_ARRAY = "0123456789abcdef".toCharArray();

    /**
     * 获取待合并的MOD所在目录
     * 这个工具默认配置的是在mods目录下
     *
     * @param meringModDir mod合并目录地址
     * @return 待合并的MOD目录路径
     */
    public static Path getMergingModDir(Path meringModDir) {
        if (meringModDir == null) {
            Path defaultPath = Path.of(userDir + File.separator + "mods");
            if (FileUtil.exists(defaultPath, true)) {
                return defaultPath;
            } else {
                throw new BusinessException(Localizations.t("TOOLS_MODS_DIR_NOT_EXIST", defaultPath));
            }
        } else {
            if (FileUtil.exists(meringModDir, true)) {
                return meringModDir;
            } else {
                throw new BusinessException(Localizations.t("TOOLS_MODS_DIR_NOT_EXIST", meringModDir));
            }
        }
    }

    /**
     * 获取待合并的MOD所在目录
     * 这个工具默认配置的是在mods目录下
     *
     * @return 待合并的MOD目录路径
     */
    public static Path getMergingModDir() {
        return getMergingModDir(null);
    }

    /**
     * 扫描指定目录中的所有文件，按扩展名过滤
     *
     * @param dir        目录路径
     * @param extensions 要查找的扩展名（如 ".pak", ".zip"）
     * @return 匹配的文件列表
     * @throws IOException 如果目录不存在或无法访问
     */
    public static List<Path> scanFiles(Path dir, String... extensions) throws IOException {
        List<Path> results = new ArrayList<>();
        if (!Files.exists(dir)) {
            throw new BusinessException("MOD合并目录mods在当前目录不存在，请创建一个mods目录");
        }
        try (Stream<Path> pathStream = Files.walk(dir)) {
            pathStream.filter(Files::isRegularFile)
                    .filter(file -> {
                        String filename = file.getFileName().toString();
                        for (String ext : extensions) {
                            if (filename.endsWith(ext)) {
                                return true;
                            }
                        }
                        return false;
                    })
                    .forEach(results::add);
        }
        return results;
    }


    public static Map<String, FileTree> indexPakFile(File file) {
        if (!file.exists()) {
            throw new BusinessException(Localizations.t("TOOLS_FILE_NOT_EXIST", file.getAbsolutePath()));
        }
        if (file.isDirectory()) {
            throw new BusinessException(Localizations.t("TOOLS_PATH_IS_DIRECTORY", file.getAbsolutePath()));
        }
        if (!StrUtil.endWithAny(file.getName(), ".pak")) {
            throw new BusinessException(Localizations.t("TOOLS_FILE_MUST_BE_PAK"));
        }
        Map<String, FileTree> pakIndexMap = new HashMap<>();
        try (ZipFile zipFile = ZipFile.builder().setFile(file).get()) {
            Enumeration<ZipArchiveEntry> entries = zipFile.getEntries();
            while (entries.hasMoreElements()) {
                ZipArchiveEntry zipEntry = entries.nextElement();
                String entryName = zipEntry.getName();
                String fileName = getEntryFileName(entryName);
                if (pakIndexMap.containsKey(fileName)) {
                    ColorPrinter.warning(Localizations.t("TOOLS_SAME_FILE_NAME_WARNING", fileName, entryName, pakIndexMap.get(fileName).getFullPathName()));
                }
                pakIndexMap.put(fileName, new FileTree(fileName, entryName, file.getName()));
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return pakIndexMap;
    }

    public static String getEntryFileName(String entryName) {
        return entryName.substring(entryName.lastIndexOf("/") + 1);
    }


    /**
     * 计算文件hash值
     */
    public static String computeHash(String content) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(content.getBytes(StandardCharsets.UTF_8));
            return bytesToHex(hash);
        } catch (NoSuchAlgorithmException e) {
            return String.valueOf(content.hashCode());
        }
    }

    private static String bytesToHex(byte[] bytes) {
        char[] hexChars = new char[bytes.length * 2];
        for (int i = 0; i < bytes.length; i++) {
            int v = bytes[i] & 0xFF;
            hexChars[i * 2] = HEX_ARRAY[v >>> 4];
            hexChars[i * 2 + 1] = HEX_ARRAY[v & 0x0F];
        }
        return new String(hexChars);
    }
}

File: src/main/java/ankol/mod/merger/tools/Tools.kt
package ankol.mod.merger.tools

import ankol.mod.merger.exception.BusinessException
import ankol.mod.merger.tools.Localizations.t
import java.nio.channels.FileChannel
import java.nio.file.Path
import java.nio.file.StandardOpenOption
import kotlin.io.path.*

object Tools {
    val userDir: String = System.getProperty("user.dir")

    val tempDir: String = System.getProperty("java.io.tmpdir")

    private val strFormatRegex = Regex("\\{}")

    private val HEX_ARRAY = "0123456789abcdef".toCharArray()

    /**
     * Get the directory containing the mods to be merged
     * This tool is configured by default to be under the mods directory
     *
     * @param meringModDir mod merging directory path, can be used to override the default merge directory
     * @return Path to the directory containing mods to be merged
     */
    fun getMergingModDir(meringModDir: Path? = null): Path {
        return if (meringModDir == null) {
            val defaultPath = Path(userDir, "mods")
            if (defaultPath.exists()) {
                defaultPath
            } else {
                throw BusinessException(t("TOOLS_DEFAULT_MODS_DIR_NOT_EXIST"))
            }
        } else {
            if (meringModDir.exists()) {
                meringModDir
            } else {
                throw BusinessException(t("TOOLS_MODS_DIR_NOT_EXIST", meringModDir))
            }
        }
    }

    /**
     * Scan all files in the specified directory, filter by extension
     *
     * @param mergedDirPath        Directory path
     * @param extensions Extensions to search for (e.g. ".pak", ".zip")
     * @return List of matching files
     */
    fun scanFiles(mergedDirPath: Path, vararg extensions: String): MutableList<Path> {
        val results = ArrayList<Path>()
        if (!mergedDirPath.exists()) {
            throw BusinessException(t("TOOLS_DEFAULT_MODS_DIR_NOT_EXIST"))
        }
        mergedDirPath.walk(PathWalkOption.FOLLOW_LINKS)
            .filter { it.isRegularFile() }
            .forEach { file: Path ->
                val filename = file.fileName.toString()
                for (ext in extensions) {
                    if (filename.endsWith(ext)) {
                        results.add(file)
                    }
                }
            }
        return results
    }

    /**
     * Get the file name in zipEntry, remove '/', return the lowercase file name
     * @return entry file name
     */
    fun getEntryFileName(entryName: String): String {
        return entryName.substring(entryName.lastIndexOf("/") + 1).lowercase()
    }

    /**
     * Recursively delete the specified path and all files and directories under it
     * @param path Path to delete
     */
    @JvmStatic
    fun deleteRecursively(path: Path) {
        if (path.notExists()) {
            return
        }
        if (path.isDirectory()) {
            path.listDirectoryEntries().forEach { child ->
                deleteRecursively(child)
            }
        }
        path.deleteIfExists()
    }

    /**
     * Copy file using zero-copy method for efficiency
     * @param sourcePath Source file path
     * @param targetPath Target file path
     * @param createParentDirs Whether to automatically create parent directories, default is true
     */
    fun zeroCopy(sourcePath: Path, targetPath: Path, createParentDirs: Boolean = true) {
        if (createParentDirs) {
            targetPath.parent?.createDirectories()
        }
        FileChannel.open(sourcePath).use { sourceChannel ->
            FileChannel.open(targetPath, StandardOpenOption.CREATE, StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING).use { targetChannel ->
                var position = 0L
                val size = sourceChannel.size()
                while (position < size) {
                    // transferTo may not transfer all data at once, need to loop
                    position += sourceChannel.transferTo(position, size - position, targetChannel)
                }
            }
        }
    }

    /**
     * Format string, replace {} placeholders with parameter values
     * @param template Template string, e.g. "Hello {} World {}"
     * @param args Parameter list
     * @return Formatted string
     */
    @JvmStatic
    fun format(template: String, vararg args: Any?): String {
        if (args.isEmpty()) return template
        var index = 0
        return template.replace(strFormatRegex) {
            if (index < args.size) args[index++].toString() else "{}"
        }
    }

    /**
     * Convert byte array to hexadecimal string
     *
     * @param bytes Byte array
     * @return Hexadecimal string
     */
    fun bytesToHex(bytes: ByteArray): String {
        val hexChars = CharArray(bytes.size * 2)
        for (i in bytes.indices) {
            val v = bytes[i].toInt() and 0xFF
            hexChars[i * 2] = HEX_ARRAY[v ushr 4]
            hexChars[i * 2 + 1] = HEX_ARRAY[v and 0x0F]
        }
        return String(hexChars)
    }
}
